src/core/genfb.h

```diff
@@ -368,98 +413,20 @@ void CGenFunctionBlock<T>::fillDataPointSpec(const CIEC_ANY &paValue, CStringDic
   }
 }
 
-template<class T>
-size_t CGenFunctionBlock<T>::calculateFBConnDataSize(const SFBInterfaceSpec &paInterfaceSpec) {
-  return sizeof(CEventConnection) * paInterfaceSpec.mNumEOs + sizeof(CDataConnection *) * paInterfaceSpec.mNumDIs +
-         sizeof(CGenDataConnection) * paInterfaceSpec.mNumDOs;
-}
-
-template<class T>
-size_t CGenFunctionBlock<T>::calculateFBVarsDataSize(const SFBInterfaceSpec &paInterfaceSpec) {
-  size_t result = 0;
-
-  result += paInterfaceSpec.mNumDIs * sizeof(CIEC_ANY *);
-  const CStringDictionary::TStringId *dataTypeIds = paInterfaceSpec.mDIDataTypeNames;
-  for (TPortId i = 0; i < paInterfaceSpec.mNumDIs; ++i) {
-    result += CTypeLib::getDataPointSize(dataTypeIds);
-  }
-
-  result += paInterfaceSpec.mNumDOs * sizeof(CIEC_ANY *);
-  dataTypeIds = paInterfaceSpec.mDODataTypeNames;
-  for (TPortId i = 0; i < paInterfaceSpec.mNumDOs; ++i) {
-    result += CTypeLib::getDataPointSize(dataTypeIds) * 2; // * 2 for connection buffer value
-  }
-
-  result += paInterfaceSpec.mNumAdapters * sizeof(TAdapterPtr);
-  return result;
-}
-
 template<class T>
 void CGenFunctionBlock<T>::setupFBInterface() {
-  freeFBInterfaceData();
 
-  const SFBInterfaceSpec &interfaceSpec = T::getFBInterfaceSpec();
-  size_t connDataSize = calculateFBConnDataSize(interfaceSpec);
-  size_t varsDataSize = calculateFBVarsDataSize(interfaceSpec);
-  mFBConnData = connDataSize ? operator new(connDataSize) : nullptr;
-  mFBVarsData = varsDataSize ? operator new(varsDataSize) : nullptr;
+  createGenEOCons();
```

Comment:
is there a special reason to have all these separated functions? couldn't you just have one and the subclass takes care then of doing what it needs?

at least in `commfb` `createGenInputData` and `createGenOutputData` it seems that one function would be enough

Reply:
Yes in many subclasses only one of the functions need to be override. In some even none. Also there is a required order of these methods. With this template method pattern subclasses are forced into a certain structre.